pkg/attestation/vtpm/vtpm.go

```diff
@@ -0,0 +1,307 @@
+
+	tw.ReadWriteCloser, err = tpm2.OpenTPM("/dev/tpmrm0")
+	if os.IsNotExist(err) {
+		tw.ReadWriteCloser, err = tpm2.OpenTPM("/dev/tpm0")
+	}
+
+	return tw, err
+}
+
+func ExtendPCR(pcrIndex int, value []byte) error {
+	rwc, err := OpenTpm()
+	if err != nil {
+		return err
+	}
+	defer rwc.Close()
+
+	fixedSha256Hash := sha3.Sum256(value)
+	if err := tpm2.PCRExtend(rwc, tpmutil.Handle(pcrIndex), tpm2.AlgSHA256, fixedSha256Hash[:], ""); err != nil {
+		return err
+	}
+
+	fixedSha384Hash := sha3.Sum384(value)
+	if err := tpm2.PCRExtend(rwc, tpmutil.Handle(pcrIndex), tpm2.AlgSHA384, fixedSha384Hash[:], ""); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func Attest(teeNonce []byte, vTPMNonce []byte, teeAttestaion bool) ([]byte, error) {
+	attestation, err := fetchVTPMQuote(vTPMNonce)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	if teeAttestaion {
+		attestation, err = addTEEAttestation(attestation, teeNonce)
+		if err != nil {
+			return []byte{}, err
+		}
+	}
+
+	return marshalQuote(attestation)
+}
+
+func FetchATLSQuote(pubKey, teeNonce, vTPMNonce []byte) ([]byte, error) {
+	attestation, err := fetchVTPMQuote(vTPMNonce)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	reportData, err := createTEEAttestationReportNonce(pubKey, attestation.GetAkPub(), teeNonce)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	attestation, err = addTEEAttestation(attestation, reportData)
+	if err != nil {
+		return []byte{}, err
+	}
+
+	return marshalQuote(attestation)
+}
+
+func VTPMVerify(quote []byte, pubKeyTLS []byte, teeNonce []byte, vtpmNonce []byte) error {
+	attestation := &attest.Attestation{}
+
+	err := proto.Unmarshal(quote, attestation)
+	if err != nil {
+		return fmt.Errorf("fail to unmarshal quote: %v", err)
+	}
+
+	ak := attestation.GetAkPub()
+	pub, err := tpm2.DecodePublic(ak)
+	if err != nil {
+		return err
+	}
+
+	cryptoPub, err := pub.Key()
+	if err != nil {
+		return err
+	}
+
+	reportData, err := createTEEAttestationReportNonce(pubKeyTLS, ak, teeNonce)
+	if err != nil {
+		return fmt.Errorf("fail to calculate report data: %v", err)
+	}
+
+	if err := quoteprovider.VerifyAttestationReportTLS(attestation.GetSevSnpAttestation(), reportData); err != nil {
+		return fmt.Errorf("failed to verify TEE attestation report: %v", err)
+	}
+
+	_, err = server.VerifyAttestation(attestation, server.VerifyOpts{Nonce: vtpmNonce, TrustedAKs: []crypto.PublicKey{cryptoPub}})
+	if err != nil {
+		return fmt.Errorf("verifying attestation: %w", err)
+	}
+
+	s256, s384 := calculatePCRTLSKey(pubKeyTLS)
+
+	if err := checkExpectedPCRValues(attestation, s256, s384); err != nil {
+		return fmt.Errorf("PCR values do not match expected PCR values: %w", err)
+	}
+
+	return nil
+}
+
+func publicKeyToBytes(pubKey interface{}) ([]byte, error) {
+	derBytes, err := x509.MarshalPKIXPublicKey(pubKey)
+	if err != nil {
+		return nil, err
+	}
+	return derBytes, nil
+}
+
+func createTEEAttestationReportNonce(pubKeyTLS []byte, ak []byte, nonce []byte) ([]byte, error) {
```

Comment:
Rename to `teeAttestReportNonce`. Let's use `attest` as the abbreviation for _attestations_ all over the codebase for the sake of simplicity.